### DISCUSSION TITLE
Best practice for passing RNGs to flax model

### DISCUSSION THREAD

**Discussion by JyChang012** [google/flax#2768] [Q&A]
Suppose I have the following module dependency, 
```
top_module -> module2 -> ...-> module{i} -> ... -> module{k}
```
where each module{i} can be selected from a growing set of candidates.
If each candidate require different RNG keys, I believe those (name, key) pairs of module{i} need to be passed to `top_module.apply`.
What I want is to have some reusable training pipeline code that can apply to any candidates actually used.
Is there a way for a module to know the rng key names needed by its descendants at run time? What is the best practice?

**Comment by cgarciae** [comment]
Hey @JyChang012, can you give a concrete example? Do you need to control (e.g. freeze) the RNG state in a specific way for subsets of Modules?

**Reply by JyChang012** [reply]
Basically I have the API code for
- An abstract top level module (subclass of `nn.Module`), that defines the input and output format of `__call__`
- Iterator that generate inputs to the abstract module
- Train loop method that accepts top level module and iterator as parameters (with other parameters like optimizer, variables, etc)
- Modules that developers can use as components to implement top level module. They might have custom RNG key name like `skip_layer` that randomly skip a layer. Some of them could have others as sub-module.

Developers can implement the top level module using component modules we provide. When top level module is passed into `Trainer`, we have no idea what RNG keys it requires. Even the developers themselves would not know especially when RNGs are not in the direct child modules.

As for the question
> Do you need to control (e.g. freeze) the RNG state in a specific way for subsets of Modules?
Not sure but probably no for now. 

Thanks!
